### PR TITLE
Fix: No token (query) or authclient (session) found

### PR DESCRIPTION
**Pull request by ArchBlood**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `develop` branch, _not_ the `master` branch if no hotfix
- [x] When resolving a specific issue, it's referenced in the PR's description (e.g. `Fix #xxx[,#xxx]`, where "xxx" is the Github issue number)
- [x] All tests are passing
- [ ] New/updated tests are included
- [x] Changelog was modified

If adding a **new feature**, the PR's description includes:
- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
Registering with an OAuth client can lead to token session issues, this addresses the issue.
fixes https://github.com/humhub-contrib/auth-facebook/issues/4

### CODE REVIEW

**Comment by martincodeinc** [comment]
Good morning .

When using the updated registrationcontroller i get the following error message after submitting the registration form. This is with a normal registration through the login screen.

yii\base\UnknownPropertyException: Getting unknown property: humhub\modules\user\Module::showRegistrationForm in /var/www/vhosts/commotio.nl/protected/vendor/yiisoft/yii2/base/Component.php:154
Stack trace:
#0 /var/www/vhosts/commotio.nl/protected/vendor/yiisoft/yii2/di/ServiceLocator.php(77): yii\base\Component->__get('showRegistratio...')
#1 /var/www/vhosts/commotio.nl/protected/humhub/modules/user/controllers/RegistrationController.php(134): yii\di\ServiceLocator->__get('showRegistratio...')
#2 [internal function]: humhub\modules\user\controllers\RegistrationController->actionIndex()
#3 /var/www/vhosts/commotio.nl/protected/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#4 /var/www/vhosts/commotio.nl/protected/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#5 /var/www/vhosts/commotio.nl/protected/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller->runAction('', Array)
#6 /var/www/vhosts/commotio.nl/protected/vendor/yiisoft/yii2/web/Application.php(103): yii\base\Module->runAction('user/registrati...', Array)
#7 /var/www/vhosts/commotio.nl/protected/vendor/yiisoft/yii2/base/Application.php(384): yii\web\Application->handleRequest(Object(humhub\components\Request))
#8 /var/www/vhosts/commotio.nl/index.php(24): yii\base\Application->run()
#9 {main}


Martin.

**Comment by martincodeinc** [comment]
HumHub 1.15.4 community edition

**Comment by martincodeinc** [comment]
I guess this is a fix on your branch; mentioned code snippets do not appear in the main branch code of humhub; e.g $existingUser = User::findByEmail

**Comment by ArchBlood** [comment]
Please note that this is using the develop branch which is setup for v1.16.x

**Comment by ArchBlood** [comment]
See the following variables that were added in v1.16.x;
https://github.com/humhub/humhub/blob/d216e221ae4c288439af6794bf5bdc9e9e02173d/protected/humhub/modules/user/Module.php#L162-L194

**Comment by ArchBlood** [comment]
As for a test I've ran one personally with the latest changes in the develop branch and see no issues currently;
The issue here was the else statement was always returning `true` on https://github.com/humhub-contrib/auth-facebook/issues/4 which is indeed fixed by this P/R.

Screenshot_1


> I guess this is a fix on your branch; mentioned code snippets do not appear in the main branch code of humhub; e.g $existingUser = User::findByEmail

`$existingUser = User::findByEmail` isn't used in the core User model this was an error on my part, so I fixed this by changing it to `$existingUser = User::findIdentity($registration->getUser()->email);`.

**Comment by ArchBlood** [comment]
@yurabakhtin when you have the time can you look over the P/R in case I made any mistakes?

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [inline protected/humhub/modules/user/controllers/RegistrationController.php:101]: It seems before run the "login" action we should check for `User::STATUS_ENABLED` as it was implemented here and as it is implemented on the auth controller, so user sees a message like "Your account is not approved yet!".

**Review by yurabakhtin — CHANGES_REQUESTED** [review]

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: It's supposed to be a check to see if an existing user has a matching email then bypasses the registration process, if there is a better solution then I'm all for this.

Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: For me it looks like the `$existingUser` will be always `null` here because if for example:

- `$registration->getUser()->email` = `address@mail.com`
- `User::findIdentity('address@mail.com');` => `User::findOne(['id' => 'address@mail.com'])`

I think we should either pass `$registration->getUser()->id` instead of `email` or find user by email like `User::findOne(['email' => 'address@mail.com'])`.

Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: > For me it looks like the `$existingUser` will be always `null` here because if for example:
> 
> * `$registration->getUser()->email` = `address@mail.com`
> * `User::findIdentity('address@mail.com');` => `User::findOne(['id' => 'address@mail.com'])`
> 
> I think we should either pass `$registration->getUser()->id` instead of `email` or find user by email like `User::findOne(['email' => 'address@mail.com'])`.

Good point, I'll have to run some tests for this method.

Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: I'm currently not seeing any issues with either methods the process is the same when an OAuth client is used as well as normal registration.

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: > I mean what if a user is already registered so `$existingUser !== null` but the user is not enabled (disabled or not approved). Will be the user logged in automatically without error message like "Your account is disabled!"?

I've tested with your concerns in mind, and can say that all this does is bypass the email steps and redirects the user to the account creation steps which if user approval is enabled then the user must wait till their account is approved, if the user email already exists they're able to connect their OAuth account and user account, the only issue that I'm seeing is that the even though the email is checked correctly the user is still redirected to the account creation step, so that is something to look into. 🤔

Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: I'd like to clarify the following;

- `$existingUser` redirects to > account creation > second click of oauth client button > logs in the user
- `$existingUser !== null` redirects to > account creation > normal registration process

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: > I am wondering how user may be found by ID when email is passed as param where ID is expected.

`$registration->getUser()->email` and `$registration->getUser()->id` are properties that belong to the `User` model instance associated with the `Registration` model would be my guess?

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: > * `$existingUser` redirects to > account creation > second click of oauth client button > logs in the user

Do you mean it works as you described even it the user is not active, i.e. when `$existingUser->status !== User::STATUS_ENABLED`?
Or such disabled user cannot be logged in?

Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: > > * `$existingUser` redirects to > account creation > second click of oauth client button > logs in the user
> 
> Do you mean it works as you described even it the user is not active, i.e. when `$existingUser->status !== User::STATUS_ENABLED`? Or such disabled user cannot be logged in?

From my tests if the registered user exists then they're logged in as normal but as stated on line 114 it makes sure that the status is `STATUS_ENABLED` for autologin, if the user is a newly registered user then from my tests the account isn't created but instead takes the user to the account creation step where then if admin approval is enabled the status is not used and the message for the user to wait for approval of the account.

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: Yes, but `email` != `id`.

I see the code `User::findIdentity($registration->getUser()->email)` executes SQL query like `SELECT * FROM user WHERE id = {$registration->getUser()->email}`.

Why did you implement it as `User::findIdentity($registration->getUser()->id)`?
Or if you need to find a user by email then we should use another method to get a user by email like `User::findOne(['email' => $registration->getUser()->email])`.

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: > From my tests if the registered user exists then they're logged in as normal

Did you test if the registered user has a "Disabled" or "Unapproved" status?
I think user should be restricted for log in action with such statuses, isn't?

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: > Yes, but `email` != `id`.
> 
> I see the code `User::findIdentity($registration->getUser()->email)` executes SQL query like `SELECT * FROM user WHERE id = {$registration->getUser()->email}`.
> 
> Why did you implement it as `User::findIdentity($registration->getUser()->id)`? Or if you need to find a user by email then we should use another method to get a user by email like `User::findOne(['email' => $registration->getUser()->email])`.

I'm still not sure why both are working for me as I'm using the current develop branch as always to keep updated on latest changes, I'm sure it's not a cache issue and looking over the codebase would require too much time for me currently. 🤔

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: > > From my tests if the registered user exists then they're logged in as normal
> 
> Did you test if the registered user has a "Disabled" or "Unapproved" status? I think user should be restricted for log in action with such statuses, isn't?

From my testing the user isn't created through this method it bypasses the process of requiring the go through the email token to get to the account creation section as stated, once the account creation form is filled out you're given the following screen;

Screenshot_1

Till the account is approved the user can't login if they're a new user, if the auth client doesn't properly handle `defaultNormalizeUserAttributeMap()` then no email will be displayed when the user registers their account, here's an example of what happens with approval needed by an admin;

Screenshot_2

The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: It is strange why for you the code like `User::findOne(['id' => 'address@mail.com'])` can really find a user.
However if the method is declared as `public static function findIdentity($id)` with doc text `Finds an identity by the given ID.` then we should pass user id as param and not email.

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: > The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔

I think if user is registered but it cannot be logged in by some reason (maybe it is disabled or not approved) then user should see a message about the reason as it is displayed when user is logged in by login/pass - https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/controllers/AuthController.php#L287-L292

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:97]: > It is strange why for you the code like `User::findOne(['id' => 'address@mail.com'])` can really find a user. However if the method is declared as `public static function findIdentity($id)` with doc text `Finds an identity by the given ID.` then we should pass user id as param and not email.

I'm not sure that going by `id` would be the best approach here as `ids` are unique to the user and can be different on other platforms that the auth client is used for, using email would be better because it's something that both accounts should have in common and are more commonly allowed. The same can be said if we were to check against `username` instead of `email`, what's used on one platform can be completely different on another.

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: > > The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔
> 
> I think if user is registered but it cannot be logged in by some reason (maybe it is disabled or not approved) then user should see a message about the reason as it is displayed when user is logged in by login/pass - https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/controllers/AuthController.php#L287-L292



> > The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔
> 
> I think if user is registered but it cannot be logged in by some reason (maybe it is disabled or not approved) then user should see a message about the reason as it is displayed when user is logged in by login/pass - https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/controllers/AuthController.php#L287-L292



> > The only issue is, if the registering user tries logging in they're redirected back to the account creation page, which I've not found a way of fixing yet... 🤔
> 
> I think if user is registered but it cannot be logged in by some reason (maybe it is disabled or not approved) then user should see a message about the reason as it is displayed when user is logged in by login/pass - https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/controllers/AuthController.php#L287-L292

That is where the original issue came into play, too many else statements ended up causing the following to always return as `true`

```php
else {	
            Yii::warning('Registration failed: No token (query) or authclient (session) found!', 'user');	
            Yii::$app->session->setFlash('error', 'Registration failed.');	
            return $this->redirect(['/user/auth/login']);
}
```

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php]: If you need to find a user by email then you should use a code like `User::findOne(['email' => $registration->getUser()->email])` instead of the current `User::findIdentity($registration->getUser()->email)` because this method searches users by id.

Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php]: Updated here https://github.com/humhub/humhub/commit/4c7f67f0277c190246bd738b55765c8f6904be99

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: ok, if you think it works as expected now then I could approve it, because I cannot test it on my local server now.
I will test this later on community server because I have similar issue with my account here https://github.com/humhub/humhub-internal/issues/258.

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/RegistrationController.php]: Thank you, now the code looks correctly.

**Review by yurabakhtin — APPROVED** [review]

File: protected/humhub/modules/user/controllers/RegistrationController.php
Comment (ArchBlood) [reply protected/humhub/modules/user/controllers/RegistrationController.php:101]: > ok, if you think it works as expected now then I could approve it, because I cannot test it on my local server now. I will test this later on community server because I have similar issue with my account here https://github.com/humhub/humhub-internal/issues/258.

Maybe the issue could be more simpler than we think? Looking at how we're getting the token and how it's validated could also be the reason why the error is happening? There are so many other possibilities that could be the root cause of it;
https://github.com/humhub/humhub/blob/ef41e8ab37b5a810e96815ac590ab41a2340a840/protected/humhub/modules/user/services/InviteRegistrationService.php#L41-L44

I've tested to see if `$token` is correct with the following and it is correct but I'm not sure if the validation is correct in the if statement `$inviteRegistrationService->isValid()`;
```php
// Log the token value
$token = Yii::$app->request->get('token');
Yii::info('Token value: ' . var_export($token, true), 'debug');

if ($token) {
    $inviteRegistrationService = new InviteRegistrationService($token);
    Yii::info('Token is present, checking validity.', 'debug');
    if (!$inviteRegistrationService->isValid()) {
        Yii::error('Invalid registration token!', 'user');
        throw new HttpException(404, 'Invalid registration token!');
    }
    $inviteRegistrationService->populateRegistration($registration);
} else {
    Yii::info('Token is not present.', 'debug');
}

if (Yii::$app->session->has('authClient')) {
    $authClient = Yii::$app->session->get('authClient');
    Yii::info('AuthClient found in session: ' . var_export($authClient, true), 'debug');
    $this->handleAuthClientRegistration($authClient, $registration);
} else {
    Yii::info('AuthClient not found in session.', 'debug');
}

if (!$token && !Yii::$app->session->has('authClient')) {
    Yii::warning('Registration failed: No token (query) or authClient (session) found!', 'user');
    Yii::$app->session->setFlash('error', 'Registration failed.');
    return $this->redirect(['/user/auth/login']);
}
```

**Comment by luke-** [comment]
@ArchBlood Unfortunately, I have only now had time to look into the error in more detail.

The error `'Registration failed: No token (query) or authclient (session) found!'` could also occur in the context of AuthClients if:
- The AuthClient does not return enough attributes for a complete automatic registration (e.g. user name or surname missing) and need to provided by the user 
- and: The class of the AuthClient is not serializable  

---

If, for example, an AuthClient cannot be serialized because of an API, the following method can be implemented to remove non-serializable properties beforehand. https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/authclient/BaseClient.php#L42-L49
Ideally, of course, the AuthClient would not have to be serializable, but there is currently no other solution. 

---

Unfortunately, your current PR is difficult to review & accept because it involves major changes and the process as a whole is very complex. 

Due to the removed `else` conditions, registrations without `token` or previously set AuthClient are now also possible.

What specific problem are you trying to solve? Or what is the problem with the current implementation?

**Comment by ArchBlood** [comment]
> @ArchBlood Unfortunately, I have only now had time to look into the error in more detail.
> 
> The error `'Registration failed: No token (query) or authclient (session) found!'` could also occur in the context of AuthClients if:
> - The AuthClient does not return enough attributes for a complete automatic registration (e.g. user name or surname missing) and need to provided by the user 
> - and: The class of the AuthClient is not serializable  
> 
> ---
> 
> If, for example, an AuthClient cannot be serialized because of an API, the following method can be implemented to remove non-serializable properties beforehand. https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/authclient/BaseClient.php#L42-L49
> Ideally, of course, the AuthClient would not have to be serializable, but there is currently no other solution. 
> 
> ---
> 
> Unfortunately, your current PR is difficult to review & accept because it involves major changes and the process as a whole is very complex. 
> 
> Due to the removed `else` conditions, registrations without `token` or previously set AuthClient are now also possible.
> 
> What specific problem are you trying to solve? Or what is the problem with the current implementation?
> 

The else statement was just relocated within the `handleAuthClientRegistration()` method directly to remove the condition where it's returning `true` even though the `token` & `AuthClient` are present.

As for the possibility of the issue being because of attributes, this would require some looking into as this issue shouldn't affect authclient modules that are handling attributes correctly, and if they're not then that means all authclient modules currently in the marketplace are affected which needs placed on a priority list. :thinking:

The only reason I say that it probably isn't related to attributes is that then authclients that are still in the core would be affected as well which doesn't seem to be the case.

**Comment by luke-** [comment]
> > @ArchBlood Unfortunately, I have only now had time to look into the error in more detail.
> > The error `'Registration failed: No token (query) or authclient (session) found!'` could also occur in the context of AuthClients if:
> > 
> > * The AuthClient does not return enough attributes for a complete automatic registration (e.g. user name or surname missing) and need to provided by the user
> > * and: The class of the AuthClient is not serializable
> > 
> > 
> > If, for example, an AuthClient cannot be serialized because of an API, the following method can be implemented to remove non-serializable properties beforehand. https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/authclient/BaseClient.php#L42-L49
> > Ideally, of course, the AuthClient would not have to be serializable, but there is currently no other solution.
> > 
> > Unfortunately, your current PR is difficult to review & accept because it involves major changes and the process as a whole is very complex.
> > Due to the removed `else` conditions, registrations without `token` or previously set AuthClient are now also possible.
> > What specific problem are you trying to solve? Or what is the problem with the current implementation?
> 
> The else statement was just relocated within the `handleAuthClientRegistration()` method directly to remove the condition where it's returning `true` even though the `token` & `AuthClient` are present.
> 

I understand that the `else` has been moved, but with that it now allows registration without prior set AuthClients. 
The `else` condition prevented this case before.

We really need to understand why the AuthClient was not set before.

> As for the possibility of the issue being because of attributes, this would require some looking into as this issue shouldn't affect authclient modules that are handling attributes correctly, and if they're not then that means all authclient modules currently in the marketplace are affected which needs placed on a priority list. 🤔
> 
> The only reason I say that it probably isn't related to attributes is that then authclients that are still in the core would be affected as well which doesn't seem to be the case.

The non-serializability of AuthClient does not affect all AuthClients. The only one I am currently aware of is SAML, as it works with anonymous functions that cannot be serialized.

**Pull request closed without merging**
